[kubeadm] Add support for clusterName in config file.

karan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**: Adds a `--cluster-name` arg to kubeadm init.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
See https://github.com/kubernetes/kube-deploy/issues/636
Code inspired by #52470

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Adds --cluster-name to kubeadm init for specifying the cluster name in kubeconfig.
```

karan commented:
/retest

karan commented:
/retest

karan commented:
Tests seem to be failing with merge conflicts - my branch and this PR are up-to-date so not sure why.

```
$ co master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.

karangoel at karangoel-macbookpro in ~/.gvm/pkgsets/go1.9.2/global/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases on master (go1.9.2)
$ git sync
Already up to date.
Everything up-to-date

karangoel at karangoel-macbookpro in ~/.gvm/pkgsets/go1.9.2/global/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases on master (go1.9.2)
$ co -
Switched to branch 'kubeadm-cluster-name'

karangoel at karangoel-macbookpro in ~/.gvm/pkgsets/go1.9.2/global/src/k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases on kubeadm-cluster-name (go1.9.2)
$ git rebase origin/master
Current branch kubeadm-cluster-name is up to date.
```

karan commented:
/assign @krousey

krousey commented:
@karan 
I assume you have a remote defined called upstream that points to github.com/kubernetes/kubernetes (`git remotes list`). Rebase like this:

```console
$ git fetch upstream
$ git checkout kubeadm-cluster-name
$ git rebase upstream/master
```

karan commented:
Correct, `upstream` is k8s/k8s.

```
$ co master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.

$ git fetch upstream

$ co -
Switched to branch 'kubeadm-cluster-name'

$ git rebase upstream/master
Current branch kubeadm-cluster-name is up to date.
```

krousey commented:
I see build errors:

```
      # k8s.io/kubernetes/cmd/kubeadm/app/phases/certs
cmd/kubeadm/app/phases/certs/certs.go:353:63: not enough arguments in call to pkiutil.NewCertificateAuthority
	have ()
	want (string)
# k8s.io/kubernetes/cmd/kubeadm/test/certs
cmd/kubeadm/test/certs/util.go:31:55: not enough arguments in call to pkiutil.NewCertificateAuthority
	have ()
```

Seems like you didn't update the logic that generates ETCD certs as well.

karan commented:
With the latest changes, this builds locally now.

```
$ bazel build //cmd/kubeadm/...:all
INFO: Analysed 177 targets (1 packages loaded).
INFO: Found 177 targets...
INFO: Elapsed time: 65.023s, Critical Path: 41.15s
INFO: Build completed successfully, 22 total actions
```

karan commented:
/assign luxas

karan commented:
This PR is ready for review, the one failing test seems to be failing for other PRs as well. I believe #60589 captures that.

krousey commented:
I see that you mainly want to change the cluster context name in the kubeconfig file. This is also changing the names the certificates are signed with. I would leave the certificates alone, because the cluster name chosen could present some sensitive material that could be publically discoverable.

Like I could create a cluster name "super-secret-project" which is fine to have as a context in a private kubeconfig file. It is not fine to have that in the certificate that anyone who attempts an handshake with the API server will see.

karan committed:
Add --cluster-name to kubeadm

karan commented:
That's good feedback Kris. I've updated the PR and excluded cert changes.

dixudx commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -184,6 +184,10 @@ func AddInitConfigFlags(flagSet *flag.FlagSet, cfg *kubeadmapiext.MasterConfigur
 		`Specify the node name.`,
 	)
+	flagSet.StringVar(
+		&cfg.ClusterName, "cluster-name", cfg.ClusterName,
```
Set `ClusterName` to `"kubernetes"` by default in L110.

```
cfg := &kubeadmapiext.MasterConfiguration{ClusterName: "kubernetes"}
```

karan replied:
Good call.

timothysc replied:
So we have a moratorium on flags, I'd be ok with a config option with defaults. 

Also, if it's part of the struct I sure hope we don't need to change so many signatures.

karan commented:
Gentle ping on the review.

dixudx commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -152,6 +152,9 @@ func AddJoinConfigFlags(flagSet *flag.FlagSet, cfg *kubeadmapiext.NodeConfigurat
 		&cfg.NodeName, "node-name", "",
 		"Specify the node name.")
+	flagSet.StringVar(
+		&cfg.NodeName, "cluster-name", "",
```
We should have a default value here, right? Instead of current empty string, which will be misleading.

Change to `flagSet.StringVar(&cfg.NodeName, "cluster-name", cfg.NodeName, "Specify the cluster name.")`.

karan replied:
Sure. Done

dixudx commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -152,6 +152,9 @@ func AddJoinConfigFlags(flagSet *flag.FlagSet, cfg *kubeadmapiext.NodeConfigurat
 	flagSet.StringVar(
 		&cfg.NodeName, "node-name", "",
 		"Specify the node name.")
+	flagSet.StringVar(
```
Set ClusterName to "kubernetes" by default in L104.

`cfg := &kubeadmapiext. NodeConfiguration{ClusterName: "kubernetes"}`

karan reviewed (commented):
Done.

dixudx commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -107,7 +107,7 @@ var (
 // NewCmdInit returns "kubeadm init" command.
 func NewCmdInit(out io.Writer) *cobra.Command {
-	cfg := &kubeadmapiext.MasterConfiguration{}
+	cfg := &kubeadmapiext.MasterConfiguration{ClusterName: "kubernetes"}
```
use `kubeadmapiext.DefaultClusterName` instead?

karan replied:
Done

dixudx commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -101,7 +101,7 @@ var (
 // NewCmdJoin returns "kubeadm join" command.
 func NewCmdJoin(out io.Writer) *cobra.Command {
-	cfg := &kubeadmapiext.NodeConfiguration{}
+	cfg := &kubeadmapiext.NodeConfiguration{ClusterName: "kubernetes"}
```
use `kubeadmapiext.DefaultClusterName` instead?

karan replied:
Done

dixudx reviewed (changes requested):
Only 2 nits, others lgtm. It' ready to go.

luxas reviewed (commented):
What's the use-case / motivation behind this?
I'm very hesitant against adding more flags to kubeadm init, in fact I think we've declared a freeze there. I could _maybe_ see it going into the config file, but I don't want to over-engineer that either. We have phases for a reason, can the use case here be done via that API instead? In other words, is this an advanced use-case enough to justify going phases route vs the init one?

/assign @timothysc

xiangpengzhao commented on cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go:
```diff
@@ -224,6 +227,8 @@ type NodeConfiguration struct {
 	// CRISocket is used to retrieve container runtime info.
 	CRISocket string `json:"criSocket,omitempty"`
+	// ClusterName is the name for the cluster in kubeconfig.
+	ClusterName string `json:"clusterName"`
```
It should be an optional field, right?

karan replied:
Right. made it so

xiangpengzhao commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -107,7 +107,7 @@ var (
 // NewCmdInit returns "kubeadm init" command.
 func NewCmdInit(out io.Writer) *cobra.Command {
-	cfg := &kubeadmapiext.MasterConfiguration{}
+	cfg := &kubeadmapiext.MasterConfiguration{ClusterName: kubeadmapiext.DefaultClusterName}
```
Is `ClusterName: kubeadmapiext.DefaultClusterName` needed to be specified here? I think it will be defaulted to the value by `legacyscheme.Scheme.Default`.

karan replied:
You're right. Did not know that's what legacyscheme did. Thanks

xiangpengzhao commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -101,7 +101,7 @@ var (
 // NewCmdJoin returns "kubeadm join" command.
 func NewCmdJoin(out io.Writer) *cobra.Command {
-	cfg := &kubeadmapiext.NodeConfiguration{}
+	cfg := &kubeadmapiext.NodeConfiguration{ClusterName: kubeadmapiext.DefaultClusterName}
```
same as above

karan replied:
Done

xiangpengzhao reviewed (commented):
just a couple of nits. otherwise LGTM. However, @luxas concern sounds reasonable.

karan committed:
Remove the default clustername, and make it optional in api

karan committed:
optional field removed in test

karan committed:
Remove clusterName flag, just use config file

karan commented:
This is now not using  a flag but config file.

timothysc reviewed (approved):
Thank you for scoping this down. 

/lgtm 
/approve

jorhett commented:
So I am concerned with how this PR played out. @luxas said 

> I'm very hesitant against adding more flags to kubeadm init, in fact I think we've declared a freeze there. I could _maybe_ see it going into the config file,

As of the current version of Kubernetes 10 months later, there doesn't appear to be any documentation for the config file. The only thing said here: https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#config-file is
> Caution: The config file is still considered beta and may change in future versions.

And worse, the man page which does show the config file option continues to say:
> --config string                        Path to kubeadm config file. WARNING: Usage of a configuration file is experimental.

So it this experimental, or beta? Those would indicate very different levels of support. And either way, there's no clear documentation of what the config file should look like.  Is this what it refers to? https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta1#hdr-Kubeadm_init_configuration_types

While I much prefer config files to command line arguments, and agree this was the right approach, it seems we need to button this up a bit more with consistent documentation.
